### DISCUSSION TITLE
Building an Capacitor Android App inside Docker Container keeps asking for Android SDK Build-Tools 30.0.3

### DISCUSSION THREAD

**Discussion by Christian1984** [ionic-team/capacitor#7254] [Q&A]
I am trying to build an Android Debug-APK from our Capacitor app inside a Docker container.

My image looks like this:

```
# from https://github.com/robingenz/docker-ionic-capacitor/blob/main/Dockerfile
FROM ubuntu:22.04

ARG JAVA_VERSION=17
ARG NODEJS_VERSION=20
# See https://developer.android.com/studio/index.html#command-tools
ARG ANDROID_SDK_VERSION=11076708
# See https://developer.android.com/tools/releases/build-tools
ARG ANDROID_BUILD_TOOLS_VERSION=33.0.0
# See https://developer.android.com/studio/releases/platforms
ARG ANDROID_PLATFORMS_VERSION=33
# See https://gradle.org/releases/
ARG GRADLE_VERSION=8.0.2
# See https://www.npmjs.com/package/@capacitor/cli
ARG CAPACITOR_VERSION=5.7.0

ENV DEBIAN_FRONTEND=noninteractive
ENV LANG=en_US.UTF-8

WORKDIR /tmp

RUN apt-get update -q

# General packages
RUN apt-get install -qy \
    apt-utils \
    locales \
    gnupg2 \
    build-essential \
    curl \
    usbutils \
    git \
    unzip \
    p7zip p7zip-full \
    python3 \
    openjdk-${JAVA_VERSION}-jre \
    openjdk-${JAVA_VERSION}-jdk

# Set locale
RUN locale-gen en_US.UTF-8 && update-locale

# Install Gradle
ENV GRADLE_HOME=/opt/gradle
RUN mkdir $GRADLE_HOME \
    && curl -sL https://services.gradle.org/distributions/gradle-${GRADLE_VERSION}-bin.zip -o gradle-${GRADLE_VERSION}-bin.zip \
    && unzip -d $GRADLE_HOME gradle-${GRADLE_VERSION}-bin.zip
ENV PATH=$PATH:/opt/gradle/gradle-${GRADLE_VERSION}/bin

# Install Android SDK tools
ENV ANDROID_HOME=/opt/android-sdk
RUN curl -sL https://dl.google.com/android/repository/commandlinetools-linux-${ANDROID_SDK_VERSION}_latest.zip -o commandlinetools-linux-${ANDROID_SDK_VERSION}_latest.zip \
    && unzip commandlinetools-linux-${ANDROID_SDK_VERSION}_latest.zip \
    && mkdir $ANDROID_HOME && mv cmdline-tools $ANDROID_HOME \
    && yes | $ANDROID_HOME/cmdline-tools/bin/sdkmanager --sdk_root=$ANDROID_HOME --licenses \
    && $ANDROID_HOME/cmdline-tools/bin/sdkmanager --sdk_root=$ANDROID_HOME "platform-tools" "build-tools;${ANDROID_BUILD_TOOLS_VERSION}" "platforms;android-${ANDROID_PLATFORMS_VERSION}"
ENV PATH=$PATH:${ANDROID_HOME}/cmdline-tools:${ANDROID_HOME}/platform-tools

# Install NodeJS
RUN curl -sL https://deb.nodesource.com/setup_${NODEJS_VERSION}.x | bash - \
    && apt-get update -q && apt-get install -qy nodejs
ENV NPM_CONFIG_PREFIX=${HOME}/.npm-global
ENV PATH=$PATH:${HOME}/.npm-global/bin
RUN npm i -g yarn

# Install Capacitor CLI
RUN npm install -g @capacitor/cli@${CAPACITOR_VERSION}

# Clean up
RUN apt-get autoremove -y \
    && apt-get clean -y \
    && rm -rf /var/lib/apt/lists/* \
    && rm -rf /tmp/*

RUN cap telemetry off | tee /var/log/app-build

WORKDIR /workdir
COPY code .
RUN rm -rf node_modules
RUN yarn
RUN cap sync | tee /var/log/app-build

WORKDIR /workdir/android
RUN gradle assembleDebug
```

I have upgraded our app from Capacitor 4 to 5, and thus Gradle to v8.0.2

Whenever I run the build I get

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':app:compileDebugJavaWithJavac'.
> Could not determine the dependencies of null.
   > Failed to install the following SDK components:
         build-tools;30.0.3 Android SDK Build-Tools 30.0.3
     Install the missing components using the SDK manager in Android Studio.
```

When I try to install `build-tools;30.0.3` I get

```
root@06b76259d9f4:/# $ANDROID_HOME/cmdline-tools/bin/sdkmanager --sdk_root=$ANDROID_HOME "build-tools;30.0.3"
Warning: Dependant package with key emulator not found!
Warning: Unable to compute a complete list of dependencies.
``` 

which appears to be a known bug that has been fixed with build-tools > 30.0.3 (https://github.com/741g/android-emulator-m1-preview/issues/14)

But then, how can I force the build to use other build tools than 30.0.3?

I tried setting the `buildToolsVersion` properties inside `android/app/build.gradle` and `android/capacitor-cordova-android-plugins/build.gradle` (within the `android {}`-section), but they this doesn't change anything. I also did a full text search for "30.0.3" which didn't yield any results.

Running `./gradlew assembleDebug` instead of using the globally installed `gradle` didn't change anything, either.

How can I force gradle to use a specific version of the Android Platform Tools in the context of a Capacitor app? What am I missing?

**Comment by theKHR3b** [comment]
Hello! I run into same issue and have workaround to this.
1. seems to be there is a bug in build-tools 30.0.3 or in sdk manager because u can't install v 30.0.3 without emulator.
2. I dont know how to install emulator in docker container.

What i did to fix issue is:
1. update capasitor to last 6 version. it will update android folder, and set targetSDK to 34.
2. after update it want new gradle plugin for build android app.
3. update gradle itself to 8.4 version for new version of plugin
4. On Mac m1/m2 will be error with aapt...... on qemu x64/86 in docker environment so build will fail.
5. In the end I added two lines, one in dockerfile:
FROM --platform=linux/x86_64 ubuntu:22.04 // added --platform ... 
and "runArgs": ["--platform=linux/amd64"] in devcontainer.json in vscode for run container in emulation of amd64

In the end I finished with succesfull build aab file with ./gradlew bundleRelease
Hope it will help
